### PR TITLE
Improve bf16 support

### PR DESCRIPTION
**Pull request by PawelGlomski-Intel**
### Changes ###
- [x] Combine `amp_out_dtype` and `enable_float_output` parameters
- [x] Add bf16 to dnnl_batch_dot
- [x] Add bf16 support to other dnnl ops

### CODE REVIEW

**Comment by PawelGlomski-Intel** [comment]
@mxnet-bot run ci [unix-cpu]

**Comment by PawelGlomski-Intel** [comment]
@mxnet-bot run ci [sanity]

**Comment by PawelGlomski-Intel** [comment]
@mxnet-bot run ci [all]

**Comment by PawelGlomski-Intel** [comment]
@mxnet-bot run ci [unix-cpu]

**Comment by PawelGlomski-Intel** [comment]
@mxnet-bot run ci [centos-gpu, unix-cpu, windows-gpu]

**Comment by PawelGlomski-Intel** [comment]
@mxnet-bot run ci [all]

**Comment by PawelGlomski-Intel** [comment]
@mxnet-bot run ci [centos-gpu, unix-cpu, windows-gpu]

**Review by bgawrych — APPROVED** [review]

**Comment by bgawrych** [comment]
@mxnet-bot run ci [centos-gpu, unix-cpu]

**Comment by bgawrych** [comment]
@mxnet-bot run ci [centos-gpu]

**Comment by bgawrych** [comment]
@mxnet-bot run ci [centos-gpu]

**Comment by bgawrych** [comment]
@mxnet-bot run ci [centos-gpu]

**Comment by bgawrych** [comment]
@mxnet-bot run ci [centos-gpu]

**Review by anko-intel — APPROVED** [review]

File: 3rdparty/mshadow/mshadow/base.h [hunk 1]
```diff
@@ -769,6 +769,10 @@ namespace isnan_typed {
   MSHADOW_XINLINE bool IsNan(volatile mshadow::half::half_t val) {
     return (val.half_ & (~MSHADOW_HALF_SIGN_BIT)) > MSHADOW_HALF_EXPONENT_BITS;
   }
+  template <>
+  MSHADOW_XINLINE bool IsNan(volatile mshadow::bfloat::bf16_t val) {
+    return (val.bf16_ & (~MSHADOW_BF16_SIGN_BIT)) > MSHADOW_BF16_EXPONENT_BITS;
```
[the remark this answers is not in the document]
Comment (bgawrych) [reply 3rdparty/mshadow/mshadow/base.h:774]: checking if float is NaN - https://stackoverflow.com/questions/26052640/why-does-gcc-implement-isnan-more-efficiently-for-c-cmath-than-c-math-h

I'm not sure if converting bfloat to float won't introduce additional overhead which can reduce benefit from calling float's isnan

File: 3rdparty/mshadow/mshadow/base.h [hunk 2]
```diff
@@ -409,7 +409,7 @@ struct DataType<half::half_t> {
 #endif
 #endif
 };
-template<>
+template <>
```
[the remark this answers is not in the document]
Comment (bgawrych) [reply 3rdparty/mshadow/mshadow/base.h:412]: can be done in other PR to keep this one simple to track relevant changes

File: src/common/utils.h [hunk 3]
```diff
@@ -929,11 +929,8 @@ inline mxnet::TShape CanonicalizeAxes(const mxnet::TShape& src) {
 }
 
 inline bool is_float(const int dtype) {
```
[the remark this answers is not in the document]
Comment (bgawrych) [reply src/common/utils.h:931]: can be done in other PR to keep this one simple to track relevant changes

**Pull request merged**